### DISCUSSION TITLE
Download an youtube video

### DISCUSSION THREAD

**Discussion by kseikyo** [johnlindquist/kit#852] [Share]
This script lets you download an youtube video into your Downloads folder using [youtube-dl-exec](https://www.npmjs.com/package/youtube-dl-exec).

All you have to do is grab the url and paste that into the script.

Also, if anyone knows of a better way to provide feedback other than the widgets, let me know 😄 

```javascript
// Name: download-youtube-video
// Description: Downloads an youtube video using the url
// Author: Lucas Sierota
// Twitter: @lucasemanuelss

import "@johnlindquist/kit";
import youtubeDlExec from "youtube-dl-exec";

const response = await widget(
  `
  <div class="w-full h-full text-center flex items-center justify-center">
    <h1 :class="responseClass">{{ content }}</h1>
  <div>
  `
);
async function downloadVideo() {
  try {
    const videoSrc = await arg("Video url:");
    const videoName = await arg("File name: (don't include extension)");
    const videoPath = "Downloads"

    const fileName = videoName !== "" ? videoName : videoSrc;
    // setting path to be downloaded and including file extension
    const path = home(videoPath, path.basename(fileName) + ".mp4");

    // set loading state on widget
    response.setState({
      content: `Downloading...`,
      responseClass: "text-yellow-600",
    });

    const res = await youtubeDlExec(videoSrc, { output: path });
    console.log(res);
    response.setState({
      content: `✅ ${fileName} downloaded at ${path} ✅`,
      responseClass: "text-green-600",
    });
  } catch (err) {
    console.log(err);
    if (
      err.stderr.includes("'NoneType' object has no attribute 'decompress'") ||
      err.stderr.includes("is not a valid URL")
    ) {
      response.setState({
        content: `🔴 Error: The url provided is not an youtube video 🔴`,
        responseClass: "text-red-600",
      });
    } else {
      response.setState({
        content: `🔴 Error ${err} 🔴`,
        responseClass: "w-full h-full text-red-600",
      });
    }
  }
}

await downloadVideo();

```

**Comment by johnlindquist** [comment]
Awesome, thanks for sharing ❤️

There are few approaches to showing progress/status.


## setStatus

This is still a bit of a WIP. I'm working on updating icons and adding "actions" to messages, but the core pieces are ready.

This will change the icon of the menubar:

```js
// Name: Testing Status

import "@johnlindquist/kit"

while (true) {
    let status = await arg("status", [
        "default",
        "success",
        "warn",
        "busy",
        "error",
        "paused"
    ])

    setStatus({
        status,
        message: `${status} message...`
    })
}
```

[Open testing-status in Script Kit](https://scriptkit.com/api/new?name=testing-status&url=https://gist.githubusercontent.com/johnlindquist/f043abc8e09c986ea8da506a5b272ca8/raw/237a4b5f4c2e08bf7dda14a71b8212e04bbdc512/testing-status.ts")

## console.log

This similar example will show how to grab a log from exec and display it:

https://github.com/johnlindquist/kit/discussions/718

## Set Div

```js
// Name: Example Set Div

import "@johnlindquist/kit"


div(md(`## Starting`))

await wait(1000)

setDiv(md(`## Good job waiting`))

await wait(1000)

setDiv(md(`## A little longer`))

await wait(1000)

// This "ends" the currently open prompt
// In this case, the "div"

submit()
```

[Open example-set-div in Script Kit](https://scriptkit.com/api/new?name=example-set-div&url=https://gist.githubusercontent.com/johnlindquist/8c2cafe57c13ba4afe0b1b4364de4598/raw/4d7bcf9f36cf485628a9fe17acb991985258ed1b/example-set-div.js")

## Logging Widget

There's a discussion around making logging widget over in our new Discord if you'd like to chat more about it:
https://discord.gg/ThBufKVd
